src/wayland/wl.c

```diff
@@ -236,35 +292,94 @@ static const struct wl_touch_listener touch_listener = {
+        wl_touch_release(seat->touch.wl_touch);
+        seat->touch.wl_touch = NULL;
+}
+
+static void destroy_seat(struct dunst_seat *seat) {
+        if (seat == NULL) {
+                return;
+        }
+
+        if (seat->pointer.wl_pointer != NULL) {
+                destroy_seat_pointer(seat);
+        }
+
+        if (seat->touch.wl_touch != NULL) {
+                destroy_seat_touch(seat);
+        }
+
+        if (seat->idle_timeout != NULL) {
+                org_kde_kwin_idle_timeout_release(seat->idle_timeout);
+                seat->idle_timeout = NULL;
+        }
+
+#ifdef HAVE_WL_EXT_IDLE_NOTIFY
+        if (seat->ext_idle_notification != NULL) {
+                ext_idle_notification_v1_destroy(seat->ext_idle_notification);
+                seat->ext_idle_notification = NULL;
+        }
+#endif
+
+        wl_list_remove(&seat->link);
+
+#ifdef WL_SEAT_RELEASE_SINCE_VERSION
+        if (wl_seat_get_version(seat->wl_seat) >= WL_SEAT_RELEASE_SINCE_VERSION) {
+                wl_seat_release(seat->wl_seat);
+        } else
+#endif
+        wl_seat_destroy(seat->wl_seat);
+        seat->wl_seat = NULL;
+
+        g_free(seat->name);
+        seat->name = NULL;
+
+        g_free(seat);
+}
+
 static void seat_handle_capabilities(void *data, struct wl_seat *wl_seat,
                 uint32_t capabilities) {
+        struct dunst_seat *seat = data;
 
-        if (ctx.pointer.wl_pointer != NULL) {
-                wl_pointer_release(ctx.pointer.wl_pointer);
-                ctx.pointer.wl_pointer = NULL;
-        }
         if (capabilities & WL_SEAT_CAPABILITY_POINTER) {
-                ctx.pointer.wl_pointer = wl_seat_get_pointer(wl_seat);
-                wl_pointer_add_listener(ctx.pointer.wl_pointer,
-                        &pointer_listener, ctx.seat);
-        }
-        if (ctx.touch.wl_touch != NULL) {
-                wl_touch_release(ctx.touch.wl_touch);
-                ctx.touch.wl_touch = NULL;
+                if (seat->pointer.wl_pointer == NULL) {
+                        seat->pointer.wl_pointer = wl_seat_get_pointer(wl_seat);
+                        wl_pointer_add_listener(seat->pointer.wl_pointer,
+                                &pointer_listener, seat);
+                }
+        } else if (seat->pointer.wl_pointer != NULL) {
+                destroy_seat_pointer(seat);
         }
+
         if (capabilities & WL_SEAT_CAPABILITY_TOUCH) {
-                ctx.touch.wl_touch = wl_seat_get_touch(wl_seat);
-                wl_touch_add_listener(ctx.touch.wl_touch,
-                        &touch_listener, ctx.seat);
+                if (seat->touch.wl_touch == NULL) {
+                        seat->touch.wl_touch = wl_seat_get_touch(wl_seat);
+                        wl_touch_add_listener(seat->touch.wl_touch,
+                                &touch_listener, seat);
+                }
+        } else if (seat->touch.wl_touch != NULL) {
+                destroy_seat_touch(seat);
```

Comment:
If I read this correctly, the touch input is destroyed when a second input is added. It seems to me that if a second input is added, the old one should stay. But let me know if it works differently.

Reply:
The touch input is destroyed when the touch capability disappears from the seat. If the capability does not change, the existing wl_touch will be kept. Same goes for pointer.

A seat can have only one touch (pointer, keyboard...) object and only if the corresponding capability was announced via capabilities event.

Reply:
Ah okay, makes sense